src/nu/validator/saxtree/TreeParser.java

```diff
@@ -298,4 +298,26 @@ public String getSystemId() {
             return locatorDelegate.getSystemId();
         }
     }
+
+    /**
+     * @see org.xml.sax.Locator#getSystemId()
+     */
+    public String getXMLVersion() {
+        if (locatorDelegate == null) {
+            return null;
+        } else {
+            return ((Locator2)locatorDelegate).getXMLVersion();
```

Comment:
Instead of casting, please change the type of `locatorDelegate` (and the arguments via which it is set) to `Locator2`.

Reply:
This one appears problematic to me - if I change the method signatures for (eg) startCDATA to take a Locator2 instead of Locator, this has knock-on effects throughout the entire API - I need to change every instance of Locator in Node, plus all subclasses and all methods that call into it. It become a very far-reaching change.

I have instead added an instanceof in this case, as for the other changes.

Reply:
OK. Thanks.